### PR TITLE
perf: keep form-action and transition graphs out of event-only bundles

### PR DESCRIPTION
**Pull request by trueadm**
> [!NOTE]
> **Low Risk**
> Changes are tree-shaking and compile-time emission guards with fail-closed checkpoint analysis; function form actions still depend on `setFormAction` installing the driver, which new tests pin.
> 
> **Overview**
> **Production bundles that only delegate events** no longer pull in `handleFormSubmit`, `startTransition`, and the held-transition graph. `dispatchDelegated` invokes a nullable `FORM_SUBMIT_DRIVER` that **`setFormAction` installs** when a function `$$formAction` is first wired, so specialized/static roots and click-only apps shrink by several KB gzip without changing submit behavior when a real function action exists.
> 
> **Compiled setup checkpoints** (`__s.block.pending` early return) are emitted only when static analysis decides setup **can** schedule a render-phase self-update. Passive built-in hooks and allowlisted pure globals no longer force the checkpoint; setters, sync-factory bodies, factories/initializers passed by reference, prop/custom-hook calls, and anything unclassified still do. Module binding names guard the global allowlist against shadowing.
> 
> Adds **reachability and behavior tests** (production esbuild bundles, spread-mounted form actions, memo-factory mount replay) and **audit doc** measurements for the bundle-size follow-up.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6ed5b204074f7f73e387573e74741b68ce5c735a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary
- Attributes the client bundle and compiled-output growth of #943 per top-level declaration (Vite production tree-shaking, base 33720d8ef vs main vs this branch) and removes the part that is reachability rather than fixed behavior.
- `dispatchDelegated` no longer references the form-action submit handler directly. `setFormAction`, the only writer of `$$formAction`, installs a `FORM_SUBMIT_DRIVER`, so bundles that only delegate events (compiler-specialized roots, the CLI SPA starter) stop retaining `startTransition` and the whole held-transition swap graph (about 30 functions, roughly 4.1 KB gzip).
- The compiled setup checkpoint (`if (__s.block.pending && !__s.block.crossRenderUpdate) return;`) is emitted only when setup can schedule a render-phase self-update. Passive built-in hooks and unshadowed standard globals cannot; setters, custom hooks, prop callbacks, sync factories that call setters, factories or initializers passed by reference (any callable-shaped argument to a sync-factory hook fails closed), and any unclassified call keep it.

## Why
- The audit doc recorded +3.3 KB gzip on the hook-memo bundles and +713 minified bytes on its compiled fixture. Measured in one environment, the static/specialized minimal scenarios had grown far more (+4.9 KB and +5.2 KB gzip): `dispatchDelegated → handleFormSubmit → startTransition → ensureTransitionSwapDriver` made the transition machinery reachable from any bundle with delegated events. Before #943 that path was only reachable through `setFormAction`.
- The checkpoint is not behaviorally observable (the runtime's own child-initialization guards back it; every pinned suite passes with it removed entirely), so it only needs to exist where a self-update is possible.

## Measurements
Same environment for every column: Node 24.18.0, macOS ARM64, locked native parser, `pnpm install --offline`. Base is 33720d8ef (the PR base), main is 44d50dbc0.

`node benchmarks/hook-memo/run.mjs` (inline variant, gzip bytes / minified bytes):

| Metric | Base | Main | This branch |
| --- | ---: | ---: | ---: |
| Tree-shaken bundle, gzip | 55,169 | 58,568 | 58,580 |
| Compiled fixture output, minified | 7,183 | 7,896 | 7,896 |

`node benchmarks/bundle-size/run-minimal.mjs` (gzip bytes):

| Scenario | Base | Main | This branch |
| --- | ---: | ---: | ---: |
| root-static-specialized | 18,565 | 23,499 | 19,410 |
| cli-spa-starter | 31,256 | 36,459 | 32,294 |
| root-static | 46,287 | 49,419 | 49,440 |
| hooks-state | 47,553 | 51,346 | 51,350 |
| context | 48,236 | 52,066 | 52,049 |
| hydrate-root | 52,873 | 55,935 | 55,942 |
| deferred-hydration | 68,872 | 72,722 | 72,717 |
| suspense-transition | (base fixture fails its own gate) | 61,092 | 61,084 |

`node benchmarks/codegen-size/run.mjs` (compiled corpus, minified / gzip):

| | Base | Main | This branch |
| --- | ---: | ---: | ---: |
| Corpus minified bytes | 76,534 | 78,983 | 76,553 |
| Corpus gzip bytes | 27,146 | 27,555 | 27,278 |

Attribution of the residual growth on root-static-specialized after this branch (about 850 gzip bytes): the urgent-transition render handling in `renderBlockInner`, the mount-time render-phase drain, delegated handler and submit snapshots, commit-phase error queue, Document/fragment root support, and the custom-element event checks. All are the fixed behaviors.

## What was not reduced, and why
- Hook-memo bundles and every `createRoot` scenario keep their growth. Their entry retains the descriptor renderer, whose spread and host-prop paths own function form actions, so they already reached the transition graph before #943 through `setFormAction → handleFormSubmit → startTransition`. That chain cannot be made lazy soundly: a spread or `createElement` prop decides at render time. The growth there is the new transition-hook records, queued state replay, and held-update reconciliation, which are the B-series fixes themselves.
- The hook-memo compiled fixture keeps all 12 checkpoints because every fixture component invokes an `observe` prop in setup, which is an unclassifiable call.
- `event-names.js` stays runtime data: custom-element spread/descriptor props need the exact camel-case delegated set at render time; feature detection would diverge from React on the pinned corner cases.
- `html-tree-validation.js` was already imported by the client runtime before #943 (dev-only use, tree-shaken from production output), so it is not a contributor.
- Every #943 development diagnostic named in the task is already behind `process.env.NODE_ENV !== 'production'`; none of their strings appear in the production bundles.
- The `benchmarks/bundle-size` ratio guards (`node benchmarks/bench.mjs --ratios bundle-reachability`) breach 66 times on main **and on the pre-#943 base**: the committed budgets date from #767 (2026-08-16) and that suite only runs in the weekly bench workflow. This branch improves the affected scenarios but I did not raise budgets; that needs a maintainer decision.

## Changes
- `packages/octane/src/runtime.ts`: `FORM_SUBMIT_DRIVER`, installed by `setFormAction`, invoked from `dispatchDelegated`.
- `packages/octane/src/compiler/compile.js`: `setupCanScheduleSelfUpdate` replaces the any-call rule for the setup checkpoint; `collectModuleBoundNames` guards the global allowlist against shadowing; the allowlist is an explicit set of non-invoking members (`Math.*`, `Number`/`Object`/`Array`/`Date` inspectors, `JSON.parse`/`stringify` with their callback positions checked like hook factories), so `Array.from`, `Object.groupBy`, and any callback passed by reference keep the checkpoint.
- Tests: `form-action-bundle.test.ts` (production reachability with and without a function form action, executed in JSDOM), spread-installed form action after mount in `audit-events-portals.test.ts`, memo-factory self-update replay plus checkpoint emission matrix in `audit-hook-behavior.test.ts`. Each was confirmed red with its fix reverted.
- `docs/divergence-audit-2026-09-03.md`: follow-up measurements row.
- Changeset: `octane` patch.

## Validation
- [x] `pnpm format:check`
- [x] `pnpm typecheck`
- [x] `pnpm test`: 28,733 passed, 2 failed as 5 s timeouts (`hmr.test.ts` prod `'App'` vite replacement, `compiler/volar.test.ts` spread-ref type check) during a 25-minute run under concurrent load; both files pass in isolation on this branch and on main (96/96). Locally the vitest orchestrator exhausts the default 4 GB heap on an unmodified `main` checkout too, so the run used `NODE_OPTIONS=--max-old-space-size=12288`.
- [x] targeted tests, both projects (`octane`, `octane-prod`): `audit-hook-behavior`, `audit-events-portals`, `compiler-authored-hosts`, `host-update-contract`, `audit-list-diagnostics`, `audit-hook-diagnostics`, `universal-manual-hook-context`, `plain-hook-import-alias`, `actions`, `production-error-bundle`, `form-action-bundle`
- [x] `node benchmarks/hook-memo/run.mjs` value/identity and observer controls pass

## Risk / follow-ups
- If a future path other than `setFormAction` starts writing `$$formAction`, it must also install the driver; the reachability test pins the negative case and the spread/descriptor tests pin the positive one.
- A component whose setup reaches a setter only through a route the analysis cannot see (an unknown call) still gets the checkpoint, so the failure mode of a misclassification is an omitted optimization, never a wrong render.

## Provenance

- [x] An agent produced this diff (`agent-authored`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/octane/src/compiler/compile.js [hunk 1]
```diff
@@ -4723,6 +4723,178 @@ function containsRenderCall(stmts, memoCtx = null) {
...
+			if (
+				callee?.type === 'MemberExpression' &&
+				!callee.computed &&
+				callee.object?.type === 'Identifier' &&
+				SETUP_PURE_GLOBAL_NAMESPACES.has(callee.object.name) &&
+				isGlobal(callee.object.name)
+			) {
+				walk(n.arguments, true);
+				return;
+			}
```
[the remark this answers is not in the document]
Comment (trueadm) [reply packages/octane/src/compiler/compile.js:4903]: Fixed in 2b25a6104: every argument to a sync-factory hook (useMemo, useState, useReducer, useSyncExternalStore, useOptimistic, useLinkedState) that could be callable now fails closed. Only inline function bodies, literals, non-callable operators, and deps/initial-value arrays keep the analysis open, so a factory or initializer passed by reference emits the checkpoint again. Pinned by three new `keeps the checkpoint` cases (memo factory by reference, state initializer by reference, conditional factory); the reference cases fail without the rule.

File: packages/octane/src/compiler/compile.js [hunk 2]
```diff
@@ -4723,6 +4723,203 @@ function containsRenderCall(stmts, memoCtx = null) {
...
```
[the remark this answers is not in the document]
Comment (trueadm) [reply packages/octane/src/compiler/compile.js:4903]: Fixed in 6ed5b2040 (commit "allow only non-invoking global namespace members"): the namespace allowlist is now an explicit member set — Math.*, Number/Object/Array/Date inspectors, and JSON.parse/stringify. Array.from, Object.groupBy/fromEntries and any unlisted member are unclassified again, and the JSON reviver/replacer positions go through the same fail-closed argument rule as hook factories, so a callback passed by reference keeps the checkpoint. Four new `keeps the checkpoint` cases (Array.from mapper, JSON reviver, JSON replacer, Object.groupBy) fail without the restriction.
